processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java

```diff
@@ -436,8 +436,36 @@ private SubclassMapping createSubclassMapping(SubclassMappingOptions subclassMap
         }
 
         private boolean isAbstractReturnTypeAllowed() {
-            return method.getOptions().getBeanMapping().getSubclassExhaustiveStrategy().isAbstractReturnTypeAllowed()
-                && !method.getOptions().getSubclassMappings().isEmpty();
+            return !method.getOptions().getSubclassMappings().isEmpty()
+                && ( method.getOptions().getBeanMapping().getSubclassExhaustiveStrategy().isAbstractReturnTypeAllowed()
+                    || isCorrectlySealed() );
+        }
+
+        private boolean isCorrectlySealed() {
+            try {
+                if ( method.getMappingSourceType().isSealed() ) {
+                    List<? extends TypeMirror> unusedPermittedSubclasses =
+                        new ArrayList<>( method.getMappingSourceType().getPermittedSubclasses() );
+                    method.getOptions().getSubclassMappings().forEach( subClassOption -> {
+                        unusedPermittedSubclasses
+                                                 .stream()
+                                                 .filter(
+                                                     permitted -> ctx
+                                                                     .getTypeUtils()
+                                                                     .isSameType(
+                                                                         permitted,
+                                                                         subClassOption.getSource() ) )
```

Comment:
I haven't played around with this, but would changing this check to be either the "same type" check or a recursive "isCorrectlySealed" check be enough to fully cover a sealed subtype hierarchy, rather than just the case where it's only one-level deep and only the direct subtypes are covered?